gpu4pyscf/qmmm/pbc/itrf.py

```diff
@@ -0,0 +1,1131 @@
+                    # s1rr[a,b,x,u,v] = 
+                    # \int phi_u [3/2*(r_a-Ri_a)(r_b-Ri_b)-1/2*(r-Ri)^2 delta_ab] (-\nable_x phi_v) dr
+                    s1rr_ = cp.asarray(-mol.intor('int1e_irrp', shls_slice=shlslc).
+                                reshape(3, 3, 3, -1, mol.nao))
+                    s1rr_trace = cp.einsum('aaxuv->xuv', s1rr_)
+                    s1rr_ *= 3 / 2
+                    for k in range(3):
+                        s1rr_[k,k] -= 0.5 * s1rr_trace
+                    s1rr.append(s1rr_)
+
+            for jatm in range(mol.natm):
+                p0, p1 = aoslices[jatm, 2:]
+
+                # d E_qm_pc / d Ri with fixed ewald_pot
+                qm_multipole_grad[jatm] += \
+                    cp.einsum('uv,xuv->x', dEds[p0:p1], s1[:,p0:p1]) \
+                  - cp.einsum('uv,xuv->x', dEds[:,p0:p1], s1[:,:,p0:p1])
+
+                # d E_qm_dip / d Ri
+                qm_multipole_grad[jatm] -= \
+                     cp.einsum('auv,axuv->x', dEdsr[:,p0:p1], s1r[jatm])
+                s1r_ = list()
+                for iatm in range(mol.natm):
+                    s1r_.append(s1r[iatm][...,p0:p1])
+                s1r_ = cp.concatenate(s1r_, axis=-2)
+                qm_multipole_grad[jatm] += cp.einsum('auv,axuv->x', dEdsr[...,p0:p1], s1r_)
+
+                # d E_qm_quad / d Ri
+                qm_multipole_grad[jatm] -= \
+                        cp.einsum('abuv,abxuv->x', dEdsrr[:,:,p0:p1], s1rr[jatm])
+                s1rr_ = list()
+                for iatm in range(mol.natm):
+                    s1rr_.append(s1rr[iatm][...,p0:p1])
+                s1rr_ = cp.concatenate(s1rr_, axis=-2)
+                qm_multipole_grad[jatm] += cp.einsum('abuv,abxuv->x', dEdsrr[...,p0:p1], s1rr_)
+
+            cput1 = logger.timer(self, 'grad_ewald pulay', *cput0)
+            s1 = s1r = s1rr = dEds = dEdsr = dEdsrr = None
+
+            ew_eta, ew_cut = cell.get_ewald_params()
+
+            # ---------------------------------------------- #
+            # -------- Ewald real-space gradient ----------- #
+            # ---------------------------------------------- #
+
+            Lall = cp.asarray(cell.get_lattice_Ls())
+
+            from pyscf import pbc
+            rmax_qm = max(cp.linalg.norm(qm_coords - cp.mean(qm_coords, axis=0), axis=-1))
+            qm_ewovrl_grad = cp.zeros_like(qm_coords)
+
+            def grad_Tij(R, r):
+                Rij = R
+                Tij = 1 / r
+                # Tija   = \nabla_a Tij
+                # Tijab  = \nabla_a Tijb
+                # Tijabc = \nabla_a Tijbc
+                Tija = -cp.einsum('ijx,ij->ijx', Rij, Tij**3)
+                Tijab   = 3 * cp.einsum('ija,ijb->ijab', Rij, Rij) 
+                Tijab   = cp.einsum('ijab,ij->ijab', Tijab, Tij**5)
+                Tijab  -= cp.einsum('ij,ab->ijab', Tij**3, np.eye(3))
+                Tijabc  = -15 * cp.einsum('ija,ijb,ijc->ijabc', Rij, Rij, Rij)
```

Comment:
Three-tensor contractions in cp.einsum consumes a large amount of GPU memory. Please decompose them into two tensor contractions manually, and replace them with `contract` in cupy_helper.